modules/core/include/opencv2/core/hal/cvtype.h

```diff
@@ -0,0 +1,256 @@
+#define CV_MAT_DEPTH_MASK         (CV_SANITY_DEPTH_MASK << CV_DEPTH_SHIFT)
+#define CV_MAT_CN_MASK            (CV_SANITY_CN_MASK    << CV_CN_SHIFT   )
+#define CV_MAT_TYPE_MASK          (CV_MAT_CN_MASK | CV_MAT_DEPTH_MASK)
+
+//ElemType register limits (public)
+#define CV_DEPTH_MAX              (1 << CV_DEPTH_LEN)
+#define CV_CN_MAX                 (1 << (CV_CN_BASE_LEN + CV_SANITY_CN_EXP_MASK)) /*(1 << CV_CN_LEN )*/
+
+//ElemType register deflate/inflate operations (private)
+#define __CV_CN_MAKE_EXPONENT(cn) static_cast<int>((~static_cast<int>((cn) > CV_SANITY_CN_BASE_MASK) + 1) & \
+                                      (0x45263107U >> (((((cn) & -static_cast<int>(cn)) * 0x17) >> 2) & 0x1C)) & CV_SANITY_CN_EXP_MASK)
+#define __CV_CN_EXPONENT(cn_bin)  (((cn_bin) >> CV_CN_BASE_LEN) & CV_SANITY_CN_EXP_MASK)
+#define __CV_CN_MAKE_BASE(cn)     (((cn) >> __CV_CN_MAKE_EXPONENT(cn))-1) /* do not do CV_SANITY_CN_BASE_MASK here */
+#define __CV_CN_BASE(cn_bin)      ((cn_bin) & CV_SANITY_CN_BASE_MASK)
+#define __CV_CN_DEFLATE(cn)       ((__CV_CN_MAKE_EXPONENT(cn) << CV_CN_BASE_LEN) | (__CV_CN_MAKE_BASE(cn) & CV_SANITY_CN_BASE_MASK)) /*((cn) - 1)*/
+#define __CV_CN_INFLATE(cn_bin)   ((__CV_CN_BASE(cn_bin) + 1) << __CV_CN_EXPONENT(cn_bin)) /*((cn_bin) + 1)*/
+
+//ElemType register pack/unpack operations (private)
+#define __CV_DEPTH_PACK(depth)    ((depth & CV_SANITY_DEPTH_MASK) << CV_DEPTH_SHIFT)
+#define __CV_DEPTH_UNPACK(flags)  (((flags) & CV_MAT_DEPTH_MASK)  >> CV_DEPTH_SHIFT)
+#define __CV_CN_PACK(cn)          ((__CV_CN_DEFLATE(cn) & CV_SANITY_CN_MASK) << CV_CN_SHIFT)
+#define __CV_CN_UNPACK(flags)     __CV_CN_INFLATE(((flags) & CV_MAT_CN_MASK) >> CV_CN_SHIFT)
+/*
+TODO: static assertion on bad input channel
+TODO: static assertion on cn == cn to avoid random multiple evaluation*/
+#define __CV_TYPE_PACK(depth,cn)  (__CV_DEPTH_PACK(depth) | __CV_CN_PACK(cn))
+/*TODO: Rewite __CV_TYPE_UNPACK*/
+#define __CV_TYPE_UNPACK(flags)   __CV_TYPE_PACK(__CV_DEPTH_UNPACK(flags),__CV_CN_UNPACK(flags)) /*((flags) & CV_MAT_TYPE_MASK)*/
+
+//ElemType register - user operations (public)
+#define CV_CN_ISVALID(cn)         (cn > 0 && cn <= CV_CN_MAX && __CV_CN_MAKE_BASE(cn) <= CV_SANITY_CN_BASE_MASK)
+#define CV_MAT_TYPE(flags)        __CV_TYPE_UNPACK(flags)
+#define CV_MAT_NO_TYPE(flags)     ((flags) & ~CV_MAT_TYPE_MASK)
+#define CV_MAT_CN(flags)          __CV_CN_UNPACK(flags)
+#define CV_MAT_DEPTH(flags)       __CV_DEPTH_UNPACK(flags)
+#define CV_MAKETYPE(depth,cn)     __CV_TYPE_PACK(depth,cn)
+#define CV_MAKE_TYPE              CV_MAKETYPE
+
+#define CV_8U        0
+#define CV_16U       2
+#define CV_32U       8
+#define CV_64U       9
+
+#define CV_8S        1
+#define CV_16S       3
+#define CV_32S       4
+#define CV_64S       11
+
+#define CV_16F       7
+#define CV_32F       5  /*avoid using this - order broken!*/
+#define CV_64F       6  /*avoid using this - order broken!*/
+#define CV_32F_      13 /*restore order among IEEE floats*/
```

Comment:
since the agreement is to use CV_MAX_DEPTH() instead of max() and do not compare depths directly, we don't need these new duplicates; let's just retain `CV_32F` etc.

Reply:
@vpisarev If we are going to break numeric-based user-code for multi-channels, then why not to break numeric-based user-code for single-channels as well? It is a chance to avoid all the hacky workarounds, such as `CV_MAX_DEPTH` etc.

Reply:
@alalek @vpisarev @mshabunin Kindly consider that by this PR, user code will get broken in case of hard-coded multi-channel numeric values (e.g. `8` instead of `CV_8UC2`, or `CV_32FC2 + 8` instead of `CV_32FC3`), so why not to restore the order among depth types?

If we restore it, then will only add extra two cases in breaking user code in case of hard-coded single-channel numeric values (5 instead of `CV_32F` and 6 instead of `CV_64F`).. In my opinion, this is nothing comparable to the multi-channel case, and at any rate, users who hard code numeric values are expected to be very few to begin with.

What do you think?

Reply:
Any decision about swapping CV_16F with CV_32S or CV_16U?
Refer to https://github.com/opencv/opencv/issues/12569#issuecomment-425026127 for details